File: javapatterns/src/main/java/org/iut/nantes/factories/pizzas/factorymethod/Pizzeria.java
package org.iut.nantes.factories.pizzas.factorymethod;

abstract public class Pizzeria {
    abstract public Pizza commander(String type);
}

class NantesPizzeria extends Pizzeria{
    @Override
    public Pizza commander(String type) {
        if ("royale".equals(type)) {
            return new RoyaleNante();
        } else
        if ("vegetarienne".equals(type)) {
            return new VegetarienneNantes();
        } else return null;
    }
}

class NYPizzeria extends Pizzeria{
    @Override
    public Pizza commander(String type) {
        if ("royale".equals(type)) {
            return new RoyaleNY();
        } else
        if ("vegetarienne".equals(type)) {
            return new VegetarienneNY();
        } else return null;
    }
}

class TokyoPizzeria extends Pizzeria{
    @Override
    public Pizza commander(String type) {
        if ("royale".equals(type)) {
            return new RoyaleTokyo();
        } else
        if ("vegetarienne".equals(type)) {
            return new VegetarienneTokyo();
        } else return null;
    }
}

File: kotlinpatterns/src/main/kotlin/org/iut/nantes/factories/pizzas/factorymethod/Pizzeria.kt
package org.iut.nantes.factories.pizzas.factorymethod

abstract class Pizzeria {
    abstract fun commander(type: String): Pizza?
}

class NantesPizzeria : Pizzeria() {
    override fun commander(type: String): Pizza? {
        return if ("royale" == type) {
            RoyaleNante()
        } else if ("vegetarienne" == type) {
            VegetarienneNantes()
        } else null
    }
}

class NYPizzeria : Pizzeria() {
    override fun commander(type: String): Pizza? {
        return if ("royale" == type) {
            RoyaleNY()
        } else if ("vegetarienne" == type) {
            VegetarienneNY()
        } else null
    }
}

class TokyoPizzeria : Pizzeria() {
    override fun commander(type: String): Pizza? {
        return if ("royale" == type) {
            RoyaleTokyo()
        } else if ("vegetarienne" == type) {
            VegetarienneTokyo()
        } else null
    }
}